Tab swiping: Integration of the new tab management

0nko opened the pull request:
Task/Issue URL: https://app.asana.com/0/72649045549333/1208648123066962/f

### Description

This PR includes the complete implementation of the tab-swiping functionality.

### Steps to test this PR

**Make sure you’ve completed the onboarding!**

#### New tab from long press
- [x] Long press on a tabs button
- [x] Verify a new tab is opened

#### New tab from the menu
- [x] Tap on the New tab menu item
- [x] Verify a new tab is opened

#### New tab from tab switcher
- [x] Go to tab switcher
- [x] Tap on the New tab button
- [x] Verify a new tab is opened

#### New tab recreation
- [x] Open a few tabs
- [x] Go to the tab switcher
- [x] Add a new tab
- [x] Verify a new tab is correctly displayed
- [x] Go back to the tab switcher
- [x] Close the NTP
- [x] Tap on the new tab button
- [x] Verify the new tab is correctly opened

#### Swiping position
- [x] Open at least 3 tabs
- [x] Select the 1st one in the tab switcher
- [x] Swipe to the last one
- [x] Go to tab switcher and verify the correct tab is selected
- [x] Tap on the 2nd tab
- [x] Verify the correct tab is opened

#### New link
- [x] Open any website
- [x] Long press on a link to open in new tab
- [x] Verify the link is opened in a new tab

#### New link in the background
- [x] Open any website
- [x] Long press on a link to open in background tab
- [x] Verify the link is opened in a new tab in the background

#### New link in the background
- [x] Make sure the DDG app is the default browser app
- [x] Open an app that can open links in custom tabs (i.e. Twitter, Facebook)
- [x] Open a link in a custom tab
- [x] Verify the link opens correctly
- [x] Move the tab to the app
- [x] Verify the tab is correctly displayed within the app

#### Switch to tab
- [x] Start typing a site address that’s already opened in another tab
- [x] Tap on the “Switch to tab” item
- [x] Verify the existing tab with the link is shown instead of opening a new tab

#### Changing the omnibar position
- [x] Open a few tabs
- [x] Go to settings and change the omnibar position
- [x] The app will reload (the cached sites are cleared)
- [x] Swipe through the tabs and verify the load correctly

#### Bookmarks
- [x] Load a site
- [x] Add it to the boorkmarks
- [x] Open a different tab
- [x] Go to Bookmarks and select the saved boormark
- [x] Verify the bookmark loads correctly

#### Default site
- [x] Set some default URL in Settings -> General -> Show on App Launch
- [x] Kill the app and relaunch it
- [x] Verify the default site is correctly loaded in a new tab
- [x] Kill the app and relaunch it
- [x] Verify the default site is loaded in the same tab as before (no new tab is added)

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/BrowserViewModel.kt:
```diff
@@ -292,7 +312,9 @@ class BrowserViewModel @Inject constructor(
     fun onTabSelected(tabId: String) {
         launch(dispatchers.io()) {
-            tabRepository.select(tabId)
+            if (tabId != tabRepository.getSelectedTab()?.tabId) {
```
shouldn’t this only affect the feature flag?

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/DuckDuckGoWebView.kt:
```diff
@@ -227,7 +227,9 @@ class DuckDuckGoWebView : WebView, NestedScrollingChild3 {
     @SuppressLint("ClickableViewAccessibility")
     override fun onTouchEvent(ev: MotionEvent): Boolean {
-        var returnValue = false
+        parent.requestDisallowInterceptTouchEvent(true)
```
why?

0nko replied:
Prevents the ViewPager swiping on the WebView.

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/tabs/DefaultTabManager.kt:
```diff
@@ -23,36 +23,56 @@ import com.duckduckgo.app.browser.BrowserActivity.Companion.LAUNCH_FROM_EXTERNAL
-        private const val MAX_ACTIVE_TABS = 40
-    }
-
     private val browserActivity = activity as BrowserActivity
```
why is this passing an Activity as collaborator?

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/tabs/DefaultTabManager.kt:
```diff
@@ -121,37 +155,59 @@ class RealTabManager @Inject constructor(
+            lastActiveTabs.add(tabId)
 
-        browserActivity.viewModel.onTabActivated(tab.tabId)
+            browserActivity.viewModel.onTabActivated(tabId)
```
this smells. The `TabManager` is injected in the `BrowserActivity`, but here you are calling the `ViewModel` from the `BrowserActivity` directly?

0nko replied:
I’ve refactored this.

malmstein commented on app/src/main/java/com/duckduckgo/app/browser/tabs/adapter/FragmentViewHolder.kt:
```diff
@@ -0,0 +1,46 @@
+import androidx.recyclerview.widget.RecyclerView
+
+/**
+ * [ViewHolder] implementation for handling [Fragment]s. Used in [ ].
```
missing link to `FragmentStateAdapter`

malmstein commented on app/src/main/java/com/duckduckgo/app/tabs/ui/TabSwitcherActivity.kt:
```diff
@@ -208,6 +209,13 @@ class TabSwitcherActivity : DuckDuckGoActivity(), TabSwitcherListener, Coroutine
             }
         }
 
+        lifecycleScope.launch {
```
what’s this for?

0nko replied:
This prevented deleting of the last tab but after Ana’s comment I got rid of it.

malmstein commented on app/src/main/java/com/duckduckgo/app/tabs/ui/TabSwitcherAdapter.kt:
```diff
@@ -129,6 +129,12 @@ class TabSwitcherAdapter(
     }
 
     override fun onBindViewHolder(holder: TabViewHolder, position: Int, payloads: MutableList<Any>) {
+        if (isDeletingEnabled) {
```
what’s this?

0nko replied:
Removed. See the previous comment.

malmstein commented on app/src/main/java/com/duckduckgo/app/tabs/ui/TabSwitcherViewModel.kt:
```diff
@@ -66,7 +71,12 @@ class TabSwitcherViewModel @Inject constructor(
 
     suspend fun onNewTabRequested(fromOverflowMenu: Boolean) {
-        tabRepository.add()
+        val emptyTab = tabs.value?.firstOrNull { it.url.isNullOrBlank() }?.tabId
```
shouldn’t this look at the feature flag?

0nko commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:
```diff
@@ -818,13 +818,24 @@ class BrowserTabFragment :
+            if (webView.isShown) {
+                webView.onResume()
+            } else {
+                webView.post {
```
Yeah, there was an issue where a tab was selected but the WebView was not ready yet. This fixes the issue by retrying. I’ll add a comment.

0nko commented on app/src/main/java/com/duckduckgo/app/browser/BrowserActivity.kt:
```diff
@@ -338,18 +376,38 @@ open class BrowserActivity : DuckDuckGoActivity() {
+        lifecycleScope.launch {
+            viewModel.tabs.flowWithLifecycle(lifecycle).collectLatest {
+                tabManager.onTabsUpdated(it)
+                lifecycleScope.launch {
```
A leftover from refactoring.

0nko commented:
Thanks for the review, @malmstein!

> There are lots of changes in this PR that affect the code beyond the feature flag. Everything should be self-contained. The app should behave exactly as now if the feature flag is disabled. In its current state, this is not the case.

I’ll double check this.

> There is not a single test for the TabManager. If you had some, you’d see that the way you’ve built it is not testable. A class injected into an Activity should not be calling the Activity’s ViewModel per direct reference.

I planned to add tests in a separate PR. 

Just for some background, the philosophy of the TabManager here is not to act as a BL model, but rather as a UI-layer container for the tab-related functionality. I kept the calls to the ViewModel from TabManager to limit the changes in other parts of the app that might use these methods. But after a closer look, it seems they are called only from the TabManager. I moved them and now the ViewModel is not referenced anymore. 

> I’m not sure about how you check if Onboarding is complete. Please ask @nalcalag to check it. Also, please add some tests for that.

This was actually suggested by @nalcalag [here](https://github.com/duckduckgo/Android/pull/5217#discussion_r1825871112).

I’ve addressed your comments and I will address the broken/missing tests next. In the meantime, the PR's ready for another round.

0nko commented:
@malmstein I checked the code and added the missing feature flag checks in all the places. You were right, it was difficult to keep track of what’s new and what’s old with all the changes so I added the old code back and wrapped it in a FF.

0nko commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:
```diff
@@ -863,6 +874,20 @@ class BrowserTabFragment :
+
+    @SuppressLint("ClickableViewAccessibility")
+    private fun disableSwipingOutsideTheOmnibar() {
+        newBrowserTab.newTabLayout.setOnTouchListener { v, event ->
```
It does, the call to this method is guarded.

0nko commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:
```diff
@@ -2449,7 +2474,11 @@ class BrowserTabFragment :
-            binding.focusDummy.requestFocus()
+
+            // prevent a crash when the view is not initiliazed yet
+            if (view != null) {
```
I don’t know exactly why this was happening but I’ve seen the app crash because of this. I suspect it could have something to do with the way ViewPager preloads fragments, but not sure.

0nko commented on app/src/main/java/com/duckduckgo/app/browser/BrowserActivity.kt:
```diff
@@ -536,6 +600,22 @@ open class BrowserActivity : DuckDuckGoActivity() {
+    @SuppressLint("ClickableViewAccessibility", "WrongConstant")
+    private fun initializeTabs() {
+        if (swipingTabsFeature.isEnabled) {
+            tabPager.adapter = tabManager.tabPagerAdapter
```
Moved it to onCreate().

0nko commented on app/src/main/java/com/duckduckgo/app/browser/BrowserViewModel.kt:
```diff
@@ -101,39 +109,51 @@ class BrowserViewModel @Inject constructor(
+        tabs.indexOf(selectedTab)
+    }.filterNot { it == -1 }
+
+    val isOnboardingCompleted: Flow<Boolean> = userStageStore.currentAppStage
```
This was suggested by @nalcalag: https://github.com/duckduckgo/Android/pull/5217#discussion_r1825871112
